lib/logstash/codecs/multiline.rb

```diff
@@ -165,12 +167,67 @@ def register
 
     @converter = LogStash::Util::Charset.new(@charset)
     @converter.logger = @logger
+
+    # TODO: (colin) I don't really understand this @last_seen_listener poutine. I needed to create
+    # this lamda to DRY across the close & auto_flush methods to pass the closure to the new
+    # @tokenizer which I figured needs to be flushed upon close. there does not seem to be
+    # explicit tests for this closing logic.
+    # what is not clear here is the initialization of @last_seen_listener which gets initialized
+    # in the accept method but the close method systematically call auto_flush which assumes the
+    # existence of @last_seen_listener. this whole logic is confusing and should be either made
+    # more explicit and self documenting OR documentation should be prodided.
+    @auto_flush_block = lambda do |event|
+      @last_seen_listener.process_event(event)
+    end
+
     if @auto_flush_interval
       # will start on first decode
       @auto_flush_runner = AutoFlush.new(self, @auto_flush_interval)
     end
-  end # def register
+  end
+
+  def decode(data, &block)
+    data = data + @delimiter if @line_based_input
 
+    @tokenizer.extract(data.force_encoding("ASCII-8BIT")).each do |line|
+      match_line(@converter.convert(line), &block)
+    end
+  end
+
+  def encode(event)
+    # Nothing to do.
+    @on_event.call(event, event)
+  end
+
+  # this is the termination or final flush API
+  #
+  # @param block [Proc] the closure that will be called for all events that need to be flushed.
+  def flush(&block)
+    remainder = @tokenizer.flush
+    match_line(@converter.convert(remainder), &block) unless remainder.empty?
+    flush_buffered_events(&block)
+  end
+
+  # TODO: (colin) I believe there is a problem here in calling auto_flush. auto_flush depends on
+  # @auto_flush_block which references @last_seen_listener which is only initialized in the context of
+  # IdentityMapCodec which in turn I believe cannot by assumed. the multiline codec could run without
+  # IdentityMapCodec.
+  def close
+    if auto_flush_runner.pending?
+      #will cancel task if necessary
+      auto_flush_runner.stop
+    end
+
+    remainder = @tokenizer.flush
+    match_line(@converter.convert(remainder), &@auto_flush_block) unless remainder.empty?
+
+    auto_flush
+  end
+
+  # TODO: (colin) what is the pupose of this accept method? AFAICT it is only used if this codec is used
+  # within the IdentityMapCodec. it is not clear when & in which context this method is used.
```

Comment:
Not true. It can be called directly or indirectly (via the IdentityMapCodec) by any input that has been modified to supply the listener callback object.